## 13.5 Smartere gjetting

#### Eller smartere telling

Vi ønsker en måte å finne kvadratroten på som er mer effektiv en telling eller gjetting.

#### Kan vi løse dette ved å bruke en form for kvalifisert gjetting eller smartere telling?

Vi må nok en gang se på hva vi har og hva vi vet.  

Vi vet at svaret vi er på jakt etter ligger mellom 1 og tallet. Altså i et intervall.  
Hva om vi gjetter på midten av intervallet og sjekker om det er for lite eller stort?  
- Er midten for stor kan vi fjerne den delen av intervallet som er større.
- Er midten for liten kan vi fjerne den delen av intervallet som er mindre.
- Halverer altså intervallet og gjentar
- Stopper når vi ikke kan redusere intervallet mer (husk at vi fortsatt bare jobber med heltall)

Programmeringsteknisk kan vi forenkle litt ved å bruke en tilleggsverdi i stedet for intervall. 
- Først setter vi tillegsverdien til halvparten av tallet og gjetter på denne. 
- For hver runde halverer vi tillegsverdien
- Legger den til eller trekker den fra den forrige gjetting avhengig om den var for stor eller liten.  

Dette løser også problemet med å vite når svaret er nært nok. Når tilleggsverdien er blitt ≤ 1 sier vi oss fornøyd.  

En slik punktvis framgangsmåte (oppskrift) for å løse et bestemt problem kalles for en algoritme.  

Legger inn litt utskrift for å kunne sjekke hvordan den fungerer under testing. Disse kan tas bort eller gjøres om til kommentarer når den skal brukes

In [2]:
# Kvadratroten av et tall uten å bruke sqrt()
# Heltallsutgave (utgangspunkt og resultat)
# Versjon 1 intervalldeling/tilleggshalvvering

tallet = int(input('Hvilket tall skal det finnes kvadratrot av?'))

tillegg = round(tallet/2) # vil bare jobbe med heltall
svar = tillegg # utgangspunkt

# Legger inn noe mer for å kunne sjekke hvor godt metoden fungerer
iter = 1 #  antall iterasjoner 
feil = round(tallet/svar - svar) # estimering av feilen
feilliste = [abs(feil)] # feilen ved hver iterasjon legges i en liste
print(' tillegg   svar   feil ')
print(f'{tillegg:7} {svar:7} {feil:7}')

while abs(feil) > 1: 
    tillegg = round(tillegg/2)
    if svar*svar > tallet:
        svar = svar - tillegg
    elif svar*svar < tallet:
        svar = svar + tillegg 
    feil = round(tallet/svar - svar) # estimering av feilen
    print(f'{tillegg:7} {svar:7} {feil:7}')
    iter += 1
    feilliste.append(feil) # legger feilen for denne iterasjonen til i lista
print ('Svar:', svar, 'Iterasjoner:', iter)
# print(feilliste)

Hvilket tall skal det finnes kvadratrot av? 5473


 tillegg   svar   feil 
   2736    2736   -2734
   1368    1368   -1364
    684     684    -676
    342     342    -326
    171     171    -139
     86      85     -21
     43      42      88
     22      64      22
     11      75      -2
      6      69      10
      3      72       4
      2      74       0
Svar: 74 Iterasjoner: 12


For hver iterasjon/runde skrives ut tillegget som gjøres, svaret som gjettes på og et estimat på feilen i gjettingen. Det er noen ting vi bør legge merke til i det som skrives ut.  

Til å begynne med er:
- Tillegget, svaret og den beregnede feilen ganske lik.
- Estimatet av feilen er alt for høyt.  

Fortegnet på feilen er negativ når gjettingen er for høy og positiv når den er for lav.  
For eksemplet som er vist (roten av 5473) ser vi også at feilen ikke nødvendigvis blir mindre for hver iterasjon. Det vil skje for mange tall.  
Den kan være ganske liten for så å plutselig øke kraftig. Dette skyldes at forrige gjetting tilfeldigvis var ganske bra, men at tillegget fortsatt er så stort at det vil ta oss lengre bort fra rett svar. 

#### Muligheter for forbedring

Det er alltid interessant å forsøke å forbedre en løsning. Her er det pekt på noen problemer med den løsningen vi har kommet fram til. Kan vi unngå disse?  
- Sørge for at vi kommer nærmere rett svar for hver iterasjon
    - Kan vi bruke feilen i stedet for et tillegg som halveres?
        - For eksempel halvparten av den
        - Har fortegn slik at det ikke er nødvedig å sjekke om "svaret" er for høyt eller lavt
- Bedre beregning av feilen?
- Bedre første gjetting?